### PR TITLE
[#657] Change how blocks are downloaded and processed

### LINKED ISSUE #657: Change how blocks are downloaded and scanned
Currently the SDK first downloads all the blocks and then the SDK scans those blocks. Disadvantage of this approach is that sync process requires lot of space on disk (2-10GB). Let's try different approach:

1. Download N blocks.
2. Scan N blocks.
3. Repeat.

Thanks to this approach sync process requires only space on disk to store N blocks.

### PR DESCRIPTION
**Pull request by Chlup**
Closes #657

- Now SDK downloads N downloads and scan those. And repeat until sync is done. This safes lot of space of disk used to store downloaded blocks.
- Most changes are done in `CompactBlockProcessor.processNewBlocks`.
- `SyncRanges` structure is changed a bit.
- `SyncStatus` is changed. SDK now report that it's syncing with progress. It doesn't report each phase of the sync process separately. Also appropriate notifications were updated.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

**Comment by pacu** [comment]
I changed the base so I could make some changes too if needed. I created a branch on the ZcashLightClientKit repo and pointed this PR to that base instead.

**Comment by pacu** [comment]
I'm testing this PR with the ECC Wallet by running this branch and pointing to the local branch with the changes.

https://github.com/zcash/zcash-ios-wallet/pull/318

**Review by pacu — APPROVED** [review]
Looks good to me.
Modulo the open items in the comments I left

**Review by pacu — CHANGES_REQUESTED** [review]
I'd add these tests and check the other comment about the dangling `else` statement
https://github.com/zcash/ZcashLightClientKit/pull/668#discussion_r1046470523

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift [hunk 1]
```diff
@@ -9,59 +9,118 @@
...
-                notifyProgress(.download(progress))
-            }
-            try await storage.write(blocks: buffer)
-            await blocksBufferWritten(buffer)
-            buffer.removeAll(keepingCapacity: true)
-        } catch {
-            guard let err = error as? LightWalletServiceError, case .userCancelled = err else {
-                throw error
+            } else {
+                break
```
[the remark this answers is not in the document]
Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift:111]: Changed to guard.

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift [hunk 2]
```diff
@@ -9,59 +9,118 @@
...
+             this:
+             0...99
+             100...199
+             200...299
+             300...399
+             ...
+             900...999
+             1000...1000
+             */
+            let processingLowerBound = range.lowerBound + (i * batchSize)
```
[the remark this answers is not in the document]
Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift:53]: Done

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift [hunk 3]
```diff
@@ -9,59 +9,120 @@
...
-        do {
-            targetHeightInternal = targetHeight
-            if targetHeight == nil {
-                targetHeightInternal = try await service.latestBlockHeightAsync()
-            }
-            guard let latestHeight = targetHeightInternal else {
-                throw LightWalletServiceError.generalError(message: "missing target height on compactBlockStreamDownload")
-            }
+        LoggerProxy.debug("Downloading blocks in range: \(range.lowerBound)...\(range.upperBound)")
+        for _ in stride(from: range.lowerBound, to: range.upperBound + 1, by: 1) {
```
Comment (pacu) [inline Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift:105]: This `range.upperBound + 1` fires some alarm. 

Stride docs
````
- Returns: A sequence from `start` toward, and possibly including, `end`. 
///   Each value in the sequence is separated by `stride`.
````
I think that `possibly` refers to the `by:` parameter not being a multiple of the position `end` has in the sequence. Our case, being the stride just by one the upper bound will always be in the sequence.

Why shall we go beyond the range given by the function parameter?

**Review by pacu — CHANGES_REQUESTED** [review]
I have one remaining question regarding the block stream stride

File: Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift (hunk 3 above)
Comment (Chlup) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift:105]: The doc you posted is for [stride(from:through:by:)](https://developer.apple.com/documentation/swift/stride(from:through:by:)). I am using `stride(from:to:by:)`. And for this method the [doc](https://developer.apple.com/documentation/swift/stride(from:to:by:)) says: "Returns a sequence from a starting value to, but not including, an end value, stepping by the specified amount." I think we are good here.

Comment (pacu) [reply Sources/ZcashLightClientKit/Block/Processor/CompactBlockDownload.swift:105]: oh I remember now why I didn't like the stride API 🫠

**Review by pacu — APPROVED** [review]

**Pull request merged**